addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -307,13 +306,23 @@ def _is_valid_partner(self, partner):
     def _is_valid_product(self, product):
```

Comment:
This method should not be used, it has one major flaw: it doesn't work in batch and is the main reason for c8d5ff55ffa5, I see that despite my best efforts it has kept getting used (it should be removed in master and in this stable version a comment should be added to use `_get_valid_products`.

If you want more performance gains you should try to replace all calls to this method by calls to `_get_valid_products`, don't bother trying to optimize it because it just cannot be faster than `_get_valid_products`.

Note that this also applies to several other methods in this module such as `_is_valid_partner`